AX: Ignore the aria-required if equivalent HTML attribute is present

hs85jeong opened the pull request:
#### 640ead36a4acc8ac8dbbf4542395f2a1c346283e
<pre>
AX: Ignore the aria-required if equivalent HTML attribute is present
<a href="https://bugs.webkit.org/show_bug.cgi?id=258548">https://bugs.webkit.org/show_bug.cgi?id=258548</a>
rdar://problem/111370591

Reviewed by Tyler Wilcock.

Modify to prioritize the HTML attribute when both `aria-required` and
`required` attribute are present.

* LayoutTests/accessibility/aria-required-expected.txt:
* LayoutTests/accessibility/aria-required.html:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::isRequired const):

Canonical link: <a href="https://commits.webkit.org/265552@main">https://commits.webkit.org/265552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ee2663f4353ca6debc7f67fbc7baa8bfd5ec1e1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11192 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11732 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12843 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13781 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11387 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13587 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11353 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9451 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13248 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10127 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17328 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10282 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13500 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10713 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8801 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9883 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10018 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/2685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14157 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10566 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -823,17 +823,16 @@ bool AccessibilityNodeObject::isMultiSelectable() const
-    // Explicit aria-required values should trump native required attributes.
+    Node* n = this->node();
+    if (is<HTMLFormControlElement>(n) && downcast<HTMLFormControlElement>(*n).isRequired())
+        return true;
```
I know this code existed before, but since you're fixing it anyway let's make it a bit cleaner 🙂 

It can be written more simply like so:

```
auto* formControlElement = dynamicDowncast<HTMLFormControlElement>(node());
if (formControlElement && formControlElement->isRequired())
    return true;
```

`dynamicDowncast<T>` combines `is<T>` and `downcast<T>`, returning `nullptr` if the value is not a `T`.

hs85jeong replied:
Thank you for review.
I applied it :)

twilco reviewed (approved):
Thank you!

hs85jeong committed:
AX: Ignore the aria-required if equivalent HTML attribute is present
https://bugs.webkit.org/show_bug.cgi?id=258548
rdar://problem/111370591

Reviewed by Tyler Wilcock.

Modify to prioritize the HTML attribute when both `aria-required` and
`required` attribute are present.

* LayoutTests/accessibility/aria-required-expected.txt:
* LayoutTests/accessibility/aria-required.html:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::isRequired const):

Canonical link: https://commits.webkit.org/265552@main
